node/pkg/watchers/solana/shim.go

```diff
@@ -269,17 +282,25 @@ func (s *SolanaWatcher) shimProcessRest(
 	var verifiedCoreEvent bool
 	var messageEvent *ShimMessageEventData
 	var err error
-	coreEventFound := false
 	for idx := startIdx; idx < len(innerInstructions); idx++ {
 		inst := innerInstructions[idx]
 		if inst.ProgramIDIndex == whProgramIndex {
-			if verifiedCoreEvent, err = shimVerifyCoreMessage(inst.Data); err != nil {
+			if verifiedCoreEvent {
+				return fmt.Errorf("detected multiple inner core instructions when there should not be at instruction %d, %d", outerIdx, idx)
+			}
+			foundIt, err := shimVerifyCoreMessage(inst.Data)
+			if err != nil {
 				return fmt.Errorf("failed to verify inner core instruction for shim instruction %d, %d: %w", outerIdx, idx, err)
 			}
-			alreadyProcessed.add(outerIdx, idx)
-			coreEventFound = true
+			if foundIt {
+				alreadyProcessed.add(outerIdx, idx)
```

Comment:
I think it's OK to add this the instruction to the alreadyProcessed set even if `foundIt` is false.

Reply:
I'd like to leave it the way it is. It think it's cleaner since we haven't "consumed" the event.